### ISSUE TITLE
Add download quotas

### ISSUE THREAD

**Issue by jpdillingham** [slskd/slskd#388]
#346 introduced slot and bandwidth limiting by group.  Expanding on that idea, file and overall size quotas should be added.

On a per-user basis, track the overall number of files and data downloaded, and prevent users from enqueuing more files if they have exceeded their configured quota for a given time period.  Include settings for daily and weekly limits to start, and consider monthly.

"daily" should be defined as the previous 24 hours, and "weekly" should be defined as the previous 7 days.

This feature depends on #387, as completed transfers need to be persisted so we can count them and aggregate their size.

**Comment by Octoputty** [comment]
I'm really enjoying this project but unfortunately my client is getting hammered with huge queues that I can't keep up with. This issue would really help and I believe it to be a necessary feature. I'm hoping it's added soon!

**Comment by jpdillingham** [comment]
This is available in the canary release that just went out, but I'm holding off on adding official documentation for it until I do some more testing.  The configuration looks like:

```yaml
groups:
  default:
    limits:
      queued:
        files: ~
        megabytes: ~
      daily:
        files: 33
        megabytes: 500
      weekly:
        files: ~
        megabytes: ~
```

Null values (`~`) mean no limits apply.  The default limits don't impact other user groups right now, but I'm considering adding "global" limits (similar to how speed limits work).

I'd encourage anyone that wants to test to create a new user group and opt a few people into at a time; if anything malfunctions people may wind up in your DMs.

Feedback appreciated greatly!

**Comment by gsuberland** [comment]
Would it be possible to add a third metric for failed transfers, as well as just files and megabytes? I've noticed that some people's clients tend to make a lot of connections that always instantly fail, retrying over and over, spamming lots and lots of _attempted_ downloads but rarely actually succeeding in downloading stuff. It would be helpful to be able to set a limit for such behaviour, after which requests are dropped for the specified time (day/week).

**Comment by jpdillingham** [comment]
That's a good suggestion!  Alternatively I could include both succeeded and failed transfers in the overall limits.  Maybe a ratio of good:bad transfers?

I think I could add a simple flag like `include_failed` or `include: good | bad | all`, something like that.  Under the hood it's a SQL query against transfer history so lots of options.

**Comment by gsuberland** [comment]
I think it'd make sense to keep it separate, since the threshold counts are likely to be quite different depending on people's needs.

**Comment by jpdillingham** [comment]
I added this in as a `failed` option, and the criteria is that the transfer status is `Completed, Errored` and that the exception is not null.  This works for transfers that started and ended before completion, but not for cancelled transfers, which I think is ok.

If I recall detection of cancelled transfers is a little sketchy so it may not work this way in all scenarios, and I think that's ok too.

These changes are all available in the canary, I'm going to do some more testing, write up the documentation, and then release it in the next day or two.

**Comment by mathiascode** [comment]
It would be undesirable to punish users who cancelled transfers anyway, since pausing a download appears the same on the uploader's end.

**Comment by gsuberland** [comment]
Yeah, cancelled transfers are fine, it's mostly just rapid-fire download attempts that always fail that're the problem. I'm thinking of setting it to something like 200 failed downloads before folks get sent to the naughty step.

**Comment by jpdillingham** [comment]
This was released in [0.20.0](https://github.com/slskd/slskd/releases/tag/0.20.0)
